ocaml/testsuite/tests/typing-layouts-or-null/reexport.ml

```diff
@@ -0,0 +1,234 @@
+(* TEST
+ flags = "-extension layouts_alpha";
+ expect;
+*)
+
+(* CR layouts v3: ['a or_null] can't be re-exported normally,
+   because users can't define their own [Null]-like constructors. *)
+module Or_null = struct
+  type ('a : value) t : value_or_null = 'a or_null =
+    | Null
+    | This of 'a
+end
+
+[%%expect{|
+Lines 2-4, characters 2-16:
+2 | ..type ('a : value) t : value_or_null = 'a or_null =
+3 |     | Null
+4 |     | This of 'a
+Error: The kind of type 'a or_null is value_or_null
+         because it is the primitive value_or_null type or_null.
+       But the kind of type 'a or_null must be a subkind of value
+         because of the definition of t at lines 2-4, characters 2-16.
+|}]
+
+module Or_null = struct
+  type ('a : value) t : value_or_null = 'a or_null
+end
+
+[%%expect{|
+module Or_null : sig type 'a t = 'a or_null end
+|}]
+
+(* Omitting the type representation leaves constructors unexported. *)
+
+let n = Or_null.Null
+
+[%%expect{|
+Line 1, characters 8-20:
+1 | let n = Or_null.Null
+            ^^^^^^^^^^^^
+Error: Unbound constructor Or_null.Null
+|}]
+
+let t v = Or_null.This v
+
+[%%expect{|
+Line 1, characters 10-22:
+1 | let t v = Or_null.This v
+              ^^^^^^^^^^^^
+Error: Unbound constructor Or_null.This
+|}]
+
+(* [@@or_null_reexport] re-exports those constructors. *)
+
+module Or_null = struct
+  type ('a : value) t : value_or_null = 'a or_null [@@or_null_reexport]
+end
+let n = Or_null.Null
+let t v = Or_null.This v
+
+[%%expect{|
+module Or_null : sig type 'a t = 'a or_null = Null | This of 'a end
```

Comment:
Printing out with `[@@or_null_reexport]` would be cool, but very low priority if it's at all difficult.

Reply:
Will be possibly done in later patches with `Variant_with_null`.

Reply:
OK. Maybe leave a CR here noting that this output is not what we might want. We can always delete the CR later.